src/ast.rs

```diff
@@ -89,11 +99,69 @@ impl UnionElement {
     pub fn get<'a>(&self, v: &'a Value) -> Iter<'a> {
         match self {
             UnionElement::Name(name) => Box::new(v.get(name).into_iter()),
+            UnionElement::Slice(slice) => {
+                if let Value::Array(arr) = v {
+                    let step = slice.step.unwrap_or(1);
+
+                    let len = arr.len();
+
+                    let start = slice
+                        .start
+                        .map(|s| if s < 0 { s + (len as isize) } else { s })
+                        .unwrap_or(if step > 0 { 0 } else { (len as isize) - 1 });
+
+                    let end = slice
+                        .end
+                        .map(|e| if e < 0 { e + (len as isize) } else { e })
+                        .unwrap_or(if step > 0 { len as isize } else { -1 });
+
+                    Box::new(array_slice(arr, start, end, step))
+                } else {
+                    Box::new(iter::empty())
+                }
+            }
             UnionElement::Index(num) => Box::new(v.get(abs_index(*num, v)).into_iter()),
         }
     }
 }
 
+fn array_slice(arr: &[Value], start: isize, end: isize, step: isize) -> Iter<'_> {
```

Comment:
I tried
```rust
array_slice(j.as_array().unwrap(), 0, -1, 1);
```

and function never terminates (see https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=8f0f37bffb80a1e5d4bd23137cfc7d3d)

Reply:
it's not clear to me from the spec whether a negative "end" is meant to be interpreted as `len + end` or if it just means that `i < end` is always false and thus the slice is always empty

Reply:
> I tried
> 
> ```rust
> array_slice(j.as_array().unwrap(), 0, -1, 1);
> ```
> 
> and function never terminates (see https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=8f0f37bffb80a1e5d4bd23137cfc7d3d)

Who needs fuzz testing with reviewers like @mkmik? Great catch!

Reply:
Actually, I can't drive this case externally because the caller of `array_slice` handles negative `start` and `end` values. So `array_slice` has some unstated preconditions.

Reply:
@mkmik I pushed that logic inside `array_slice` to make it total.

Reply:
This is all moot after rebasing on slyce.